api: Convert libnet's errors into HTTP status code

akerouanton opened the pull request:
**- What I did**

Both libnetwork's BadRequestError and NoServiceError interfaces used a different method name than their equivalent in `errdefs`. As the package httpstatus wasn't aware of these interfaces, any libnetwork's errors implementing those interfaces were returning a 500 and a "FIXME" mesage would be logged. This is now fixed by replacing libnetwork's interfaces with newer interfaces compatible with errdefs' ones.

**- How to verify it**

CI is green

**- A picture of a cute animal (not mandatory but encouraged)**

![](https://www.thesprucepets.com/thmb/wbacRYRNfOaMJ0aeDXtsQUxI0Rk=/5200x0/filters:no_upscale():strip_icc()/mandarinfish-swimming-in-coral-reef-166264759-5ac6c034303713003775cb26.jpg)

thaJeztah commented on api/server/httpstatus/status.go:
```diff
@@ -150,3 +155,14 @@ func statusCodeFromContainerdError(err error) int {
+
+func statusCodeFromLibnetworkError(err error) int {
+	switch {
+	case nettypes.IsBadRequestError(err):
```
can we change the libnetwork errors to match the `errdefs` signature?

akerouanton replied:
There're only 10 structs implementing `BadRequestError` and `NoServiceError` so it should be easy to do. However, could you share why you prefer that over the current implementation?

thaJeztah replied:
The idea behind the errdefs package was to define a set of interfaces/types that can be implemented by packages in our code. Using those types means that code "somewhere in the stack" can produce an error of a specific "errdefs" type, and any other code consuming it is able to use that information (e.g. to check if the error was an `errdefs.IsNotFound()`); any other error (baring some errors we'd get from Go's stdlib, or when _directly_ dealing with sentinel errors) can be considered "just an error".

I'd rather not make this package handle specific implementations of "other" type of errors beyond the types that are defined by `errdefs`. Ultimately, we should also get rid of the special cases for containerd / grpc errors and the distribution errors; we should make sure that such errors are converted to a proper `errdefs` type (where needed) before they propagate through our code, but I think at the time there were still too many paths to check in that respect, so we left that for a "later exercise"

Now that libnetwork is in this repository (no longer an external repository), and we have a better grip on what kind of errors can be returned (and when), we should update the libnetwork types to implement `errdefs` types (where needed - i.e., where their error should be considered a specific type).

akerouanton committed:
libnet: Replace BadRequest with InvalidParameter

InvalidParameter is now compatible with errdefs.InvalidParameter. Thus,
these errors will now return a 400 status code instead of a 500.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
libnet: Replace NoServiceError with UnavailableError

UnavailableError is now compatible with errdefs.UnavailableError. These
errors will now return a 503 instead of a 500.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton commented:
> I'll keep it as a draft for now to see if anything breaks and how much FIXME this eliminiates. Also, I need to check if all routes that could trigger this have 400/503 documented.

Looks like the CI never triggers one of these libnet's errors, so it's not possible to evaluate the impact of this PR.

thaJeztah commented on libnetwork/ipam/structures.go:
```diff
@@ -55,22 +55,22 @@ func (s *PoolID) String() string {
 	sub, err := netip.ParsePrefix(p[1] + "/" + p[2])
 	if err != nil {
-		return types.BadRequestErrorf("%v", err)
+		return types.InvalidParameterErrorf("%v", err)
```
😭 so... we're converting an error to a string so that we can produce an error 😞 

I guess we could change `invalidParameter` to be a struct instead of a string;






But honestly,, at that point we should start wondering if we should just use the errdefs helpers directly;


```suggestion
		return errdefs.InvalidParameter(fmt.Errorf("invalid string form for subnetkey: %s", str))
```

That way we can also Unwrap errors etc, which currently isn't possible.

thaJeztah replied:
If we want to keep the `types` helpers, we could still make the utilities a wrapper, e.g.

```go
// InvalidParameterErrorf creates an instance of InvalidParameterError
func InvalidParameterErrorf(format string, params ...interface{}) error {
	return errdefs.InvalidParameter(fmt.Errorf(format, params...))
}
```

akerouanton committed:
libnet: Fix error capitalization

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

thaJeztah reviewed (approved):
LGTM
